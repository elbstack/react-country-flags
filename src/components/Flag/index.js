import React, { PureComponent as Component, createElement } from 'react';
import PropTypes from 'prop-types';
import flags from '../../flags/react';

const fetchFlag = async country => {
  if (!flags[country]) return Promise.resolve(null);
  const { default: data } = await flags[country]();
  return data;
};

// thanks to https://github.com/facebook/react/issues/5465#issuecomment-157888325
const makeCancelable = promise => {
  let hasCanceled_ = false;

  const wrappedPromise = new Promise((resolve, reject) => {
    promise.then(
      val => (hasCanceled_ ? reject({ isCanceled: true }) : resolve(val)),
      error => (hasCanceled_ ? reject({ isCanceled: true }) : reject(error))
    );
  });

  return {
    promise: wrappedPromise,
    cancel() {
      hasCanceled_ = true;
    },
  };
};

export const countries = (() => Object.keys(flags))();

export default class Flags extends Component {
  cancelUpdate = () => {};

  state = {
    flag: null,
  };

  static defaultProps = {
    country: null,
    asSquare: false,
  };

  componentDidMount() {
    this.updateFlag();
  }

  componentDidUpdate(prevProps) {
    const { country } = this.props;

    if (country !== prevProps.country) {
      this.updateFlag();
    }
  }

  componentWillUnmount() {
    this.cancelUpdate();
  }

  updateFlag() {
    const { country } = this.props;
    const { promise, cancel } = makeCancelable(fetchFlag(country));
    promise.then(flag => {
      this.setState({ flag });
    });
    this.cancelUpdate = cancel;
  }

  render() {
    const { asSquare, country, ...props } = this.props;
    const { flag } = this.state;
    if (!flag && !asSquare) return null;
    if (!asSquare) return createElement(flag, props);
    return (
      <svg viewBox="0 0 100 100" preserveAspectRatio="xMidYMid" {...props}>
        {!flag ? null : createElement(flag)}
      </svg>
    );
  }
}

Flags.propTypes = {
  country: PropTypes.string,
  asSquare: PropTypes.bool,
};
