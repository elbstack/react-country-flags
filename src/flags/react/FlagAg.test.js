import React from 'react';
import renderer from 'react-test-renderer';
import Flag from './FlagAg';

it('FlagAg renders without crashing', () => {
  const component = renderer.create(<Flag />);

  let tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});
