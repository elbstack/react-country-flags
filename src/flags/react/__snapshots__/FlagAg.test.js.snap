// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FlagAg renders without crashing 1`] = `
<svg
  viewBox="0 0 138 92"
  xmlnsXlink="http://www.w3.org/1999/xlink"
>
  <path
    d="M0 0h138v92H0z"
    fill="#fff"
  />
  <path
    d="M0 0h138v36H0z"
  />
  <g
    transform="translate(69 36)"
  >
    <g
      id="sf44718e4"
    >
      <path
        d="M-30 0L0-5.742V5.742z"
        fill="#fcd116"
        id="sc66e9395"
      />
      <use
        transform="rotate(22.5)"
        xlinkHref="#sc66e9395"
      />
      <use
        transform="rotate(45)"
        xlinkHref="#sc66e9395"
      />
    </g>
    <use
      transform="rotate(67.5)"
      xlinkHref="#sf44718e4"
    />
    <use
      transform="rotate(135)"
      xlinkHref="#sf44718e4"
    />
  </g>
  <path
    d="M0 36h138v20H0z"
    fill="#0072c6"
  />
  <path
    d="M0 0v92h68M138 0v92H68"
    fill="#ce1126"
  />
</svg>
`;
